### ISSUE TITLE
Yielded twisted failure marked as missed

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#440]
*Originally reported by* **Jakob de Maeyer (Bitbucket: [jdemaeyer](https://bitbucket.org/jdemaeyer), GitHub: [jdemaeyer](https://github.com/jdemaeyer))**

----------------------------------------

Hi,

in [this except clause (lines 75 through 81)](https://github.com/jdemaeyer/scrapy/blob/76440e767e015d2d3bcad38fecac57ac49dec287/scrapy/crawler.py#L69), we catch all exceptions, do some cleanup, and then reraise the exception by wrapping it in a `twisted.Failure`. The whole function is wrapped inside twisted's `inlineCallbacks` magic. 

The [coverage report](https://codecov.io/github/scrapy/scrapy/scrapy/crawler.py?ref=6f10968f1a3baed94bc1d5d5419c7161f20320c3#l-69) ([XML here](http://pastebin.com/rCPWQrbZ), run with `timid = True`) shows the variable assignment in line 80 being hit but the subsequent line being missed. An [accompanying test](https://github.com/jdemaeyer/scrapy/blob/76440e767e015d2d3bcad38fecac57ac49dec287/tests/test_crawl.py#L230) passes when line 81 is there and fails when I remove it, so I'm pretty sure it should indeed be hit.

I tried to create a minimal working example [here](https://github.com/jdemaeyer/codecovtest) but failed, sorry :/

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/440

**Comment by nedbat** [comment]
Can you provide me with detailed instructions on how to run the test suite?

BTW: I used scrapy yesterday, good stuff :)

**Comment by nedbat** [comment]
*Original comment by* **Jakob de Maeyer (Bitbucket: [jdemaeyer](https://bitbucket.org/jdemaeyer), GitHub: [jdemaeyer](https://github.com/jdemaeyer))**

----------------------------------------

Hey Ned, thanks for looking into it.

I'm afraid I'm not too familiar with the test suite internals. I clone the repository, then run `tox2` in it, or `tox2 -- tests/test_crawl.py::CrawlTestCase::test_graceful_crawl_error_handling` to run the relevant test only, after enabling timid mode in `.coveragerc`.

btw I use coverage.py *every* day, also good stuff :)

**Comment by nedbat** [comment]
@jdemaeyer The line you're talking about ("testvar = True") isn't in the current version of the repo (commit [57f87b95d4d705f8afdd8fb9f7551033a7d88ee2 (bb)](https://bitbucket.org/ned/coveragepy/commits/57f87b95d4d705f8afdd8fb9f7551033a7d88ee2)).  Did you add it there specifically for investigating this bug?  Is there a problem with the code as it exists in the repo?

**Comment by nedbat** [comment]
*Original comment by* **Jakob de Maeyer (Bitbucket: [jdemaeyer](https://bitbucket.org/jdemaeyer), GitHub: [jdemaeyer](https://github.com/jdemaeyer))**

----------------------------------------

Hey Ned! Yeah, the `testvar` line was to make sure that the yield statement really is hit. The code as it is in the current master works as expected (i.e. the failure is yielded). There is a test that explicitly tests that [here](https://github.com/scrapy/scrapy/blob/master/tests/test_crawl.py#L249) but the line is [reported as missed](https://codecov.io/github/scrapy/scrapy/scrapy/crawler.py?ref=54216d7afe9d545031c57b5821f2c821faa2ccc3#l-80).

**Comment by nedbat** [comment]
@jdemaeyer Looks like this is a really simple thing that doesn't involve Twisted at all.  I didn't understand your smaller test cases that didn't work.  I think this boils down to: yields that are never returned to are marked as not executed:
```
$ cat foo.py
def gen():
    print("yup")
    yield "yielded"
    print("nope")

print(next(gen()))
$ coverage run --branch foo.py
yup
yielded
$ coverage report -m
Name     Stmts   Miss Branch BrPart  Cover   Missing
----------------------------------------------------
foo.py       5      2      0      0    60%   3-4
```
When run without branch coverage, it's understood better:
```
$ coverage run foo.py
yup
yielded
$ coverage report -m
Name     Stmts   Miss  Cover   Missing
--------------------------------------
foo.py       5      1    80%   4
```

**Comment by nedbat** [comment]
Fixed in [b03f78976a78 (bb)](https://bitbucket.org/ned/coveragepy/commits/b03f78976a78)

**Comment by nedbat** [comment]
Issue #482 was marked as a duplicate of this issue.
